torch/distributed/_tensor/experimental/distribute_function.py

```diff
@@ -0,0 +1,90 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+
+import collections
+import importlib
+import types
+from typing import Callable, Dict, List, Optional, Set, Tuple
+
+import torch
+from torch import nn
+from torch.distributed.device_mesh import DeviceMesh
+
+
+_original_functions: Dict[Callable, Callable] = {}
+_wrapper_functions: Dict[Callable, Callable] = {}
+_replaced_objs: collections.defaultdict[
+    Callable, Set[Tuple[types.ModuleType, str]]
+] = collections.defaultdict(set)
+
+
+def distribute_function(
```

Comment:
I'm thinking that we should probably simplify this to be sth just a function wrapper with input_placements and output_placements similar to the `local_map` variant? https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/experimental/local_map.py#L24-L25

basically we can hide input_fn and output_fn to be inside the function wrapper, by just creating DTensors for args according to the input_placements. For out_placements, we can run sharding propagation inside the distributed function (where the output tensors should be DTensors), and the out_placements would be a user hint, where if user provide it, we should check if the output placements be the same as the output DTensor, if not we redistribute to the user specified output placements